### PR TITLE
Remove warnings discovered with -Wshadow.

### PR DESCRIPTION
**Pull request by rec**
@nbougalis @scottschurr 

Unfortunately, we can't yet turn on -Wshadow.

Some of the third party libraries have issues - we could fix those in the SConstruct - but boost has a warning that appears several times, even in 1.58.  I'll look at how I might contribute to the boost repo to get rid of this warning...

### CODE REVIEW

File: src/beast/beast/module/core/native/linux_Files.cpp [hunk 1]
```diff
@@ -201,8 +200,8 @@ class DirectoryIterator::NativeIterator::Pimpl
     DIR* dir;
 };
 
-DirectoryIterator::NativeIterator::NativeIterator (const File& directory, const String& wildCard)
-    : pimpl (new DirectoryIterator::NativeIterator::Pimpl (directory, wildCard))
+DirectoryIterator::NativeIterator::NativeIterator (const File& directory, const String& wildCard_)
+    : pimpl (new DirectoryIterator::NativeIterator::Pimpl (directory, wildCard_))
```
Comment (scottschurr) [inline src/beast/beast/module/core/native/linux_Files.cpp]: A slight grimace here.  I generally expect variables with a trailing `_` to be members.  Perhaps `wildCardIn` would not bring the wrath of the character counters?

File: src/ripple/app/ledger/impl/LedgerConsensus.cpp [hunk 2]
```diff
@@ -401,10 +401,10 @@ class LedgerConsensusImp
 
         // Adjust tracking for each peer that takes this position
         std::vector<NodeID> peers;
-        for (auto& it : mPeerPositions)
+        for (auto& i : mPeerPositions)
```
Comment (scottschurr) [inline src/ripple/app/ledger/impl/LedgerConsensus.cpp]: Nice change, since `i` is no longer an iterator.  But `i` doesn't seem quite right to me either.  Perhaps `peerPos`?

File: src/ripple/app/ledger/impl/LedgerConsensus.cpp [hunk 3]
```diff
@@ -1453,11 +1453,11 @@ class LedgerConsensusImp
         }
 
         // Update votes on disputed transactions
-        for (auto& it : mDisputes)
+        for (auto& dispute : mDisputes)
```
Comment (scottschurr) [inline src/ripple/app/ledger/impl/LedgerConsensus.cpp]: Nice!

File: src/ripple/app/ledger/impl/LedgerConsensus.cpp [hunk 4]
```diff
@@ -1522,20 +1522,20 @@ class LedgerConsensusImp
                 << mPeerPositions.size () << " nw:" << neededWeight
                 << " thrV:" << threshVote << " thrC:" << threshConsensus;
 
-            for (auto it = closeTimes.begin ()
-                , end = closeTimes.end (); it != end; ++it)
+            for (auto i = closeTimes.begin ()
+                , end = closeTimes.end (); i != end; ++it)
```
Comment (scottschurr) [inline src/ripple/app/ledger/impl/LedgerConsensus.cpp]: Hmmm.  `i` may be the best we can do here, since there's already a `closeTime` in scope.

**Comment by scottschurr** [comment]
Thanks for making this change.  Even if we can't turn on -Wshadow you have given it less reason to complain if we could have it.

There is one part that makes me wince, however.  There are several places where you've appended a trailing underscore (`_`) to parameter names.  Yup, that avoids the shadowing.  But for those of us who have been trained to recognize the trailing underscore as an indicator of a member variable it might promote some confusion.  You can try to convince me that I should change my expectations.  Or we could find an alternative to the cases where you've used trailing underscores.  Either path works for me.

**Comment by rec** [comment]
Oh, I should have commented, Scott, that I got rid of the trailing underscores in constructor parameters.

As for the iterators, I changed most of them but in one case I left it as `i` because all the good names were already in scope and because `i` is pretty clearly an iterator.

File: src/beast/beast/module/core/native/linux_Files.cpp [hunk 5]
```diff
@@ -145,9 +144,9 @@ File File::getSpecialLocation (const SpecialLocationType type)
 class DirectoryIterator::NativeIterator::Pimpl
 {
 public:
-    Pimpl (const File& directory, const String& wildCard_)
+    Pimpl (const File& directory, const String& _wildCard)
```
Comment (scottschurr) [inline src/beast/beast/module/core/native/linux_Files.cpp]: Jeepers!  I know that the standard only reserves names with leading underscores if they are followed by an upper case letter of another underscore (http://stackoverflow.com/questions/228783/what-are-the-rules-about-using-an-underscore-in-a-c-identifier).

However my habit for years has been to never name anything with a leading underscore.  Sutter and Alexandrescu agree with my take in **C++ Coding Standards** *Item 0*: "a) never use 'underhanded names,' ones that begin with an underscore or contain a double underscore;...".  So with this change I just went from uncomfortable to more uncomfortable.  But I can't say that it violates the standard.

Comment (vinniefalco) [inline src/beast/beast/module/core/native/linux_Files.cpp]: There should be no controversy here, its addressed in our coding style:
http://i.imgur.com/Z7MSpl1.png

File: src/ripple/basics/base_uint.h [hunk 6]
```diff
@@ -228,12 +228,17 @@ class base_uint
         return *this;
     }
 
+    // be32toh and htobe32 are macros that somehow cause shadowing
+    // warnings in this header file, so we hide them...
+    static uint32_t bigendToHost (uint32_t x) { return be32toh(x); }
+    static uint32_t hostToBigend (uint32_t x) { return htobe32(x); }
```
Comment (scottschurr) [inline src/ripple/basics/base_uint.h]: Good names here.  Thanks.  I really had no clue what those macro names meant.

**Comment by vinniefalco** [comment]
I've reviewed all these changes and come to this conclusion:

Compiler implementations of shadow warnings are overly aggressive and creates many false positives (for example, in a lambda with no captures). A consequence of "fixing" this broken compiler behavior is a violation of our naming convention (quoted at the end). I don't think we can sanely enable shadow warnings for the primary build. I think its okay doing a local build with the warning enabled from time to time with the intent of fixing actual bugs. I don't think this pull request should be merged.

:-1:

Coding Style:
1. Non-public data members have a trailing underscore
2. Public data members do not have a trailing underscore
3. Formal arguments used to initialize public data members in constructor initializer lists have a trailing underscore.

File: src/ripple/app/ledger/impl/LedgerConsensus.cpp (hunk 4 above)
Comment (vinniefalco) [inline src/ripple/app/ledger/impl/LedgerConsensus.cpp]: @HowardHinnant Is there any benefit to caching the return value of `end()`?

File: src/beast/beast/module/core/native/linux_Files.cpp [hunk 7]
```diff
@@ -201,8 +200,8 @@ class DirectoryIterator::NativeIterator::Pimpl
     DIR* dir;
 };
 
-DirectoryIterator::NativeIterator::NativeIterator (const File& directory, const String& wildCard)
-    : pimpl (new DirectoryIterator::NativeIterator::Pimpl (directory, wildCard))
+DirectoryIterator::NativeIterator::NativeIterator (const File& directory, const String& wildCard_)
```
Comment (scottschurr) [inline src/beast/beast/module/core/native/linux_Files.cpp]: There's still a parameter name with a trailing underscore here on line 203.  But also note my discomfort with names with a leading underscore.

File: src/ripple/nodestore/tests/Timing.test.cpp [hunk 8]
```diff
@@ -342,12 +342,12 @@ class Timing_test : public beast::unit_test::suite
 
         public:
             Body (std::size_t id, suite& s,
-                    Params const& params, Backend& backend)
+                    Params const& params_, Backend& be)
```
Comment (scottschurr) [inline src/ripple/nodestore/tests/Timing.test.cpp]: There's still a parameter name with a trailing underscore here on line 345.

File: src/ripple/overlay/impl/Manifest.h [hunk 9]
```diff
@@ -164,10 +164,10 @@ class ManifestCache
         MappedType(MappedType&&) = default;
         MappedType& operator=(MappedType&&) = default;
 #endif
-        MappedType(std::string comment,
+        MappedType(std::string comment_,
```
Comment (scottschurr) [inline src/ripple/overlay/impl/Manifest.h]: There's still a parameter name with a trailing underscore here on line 167.

File: src/ripple/app/ledger/impl/LedgerConsensus.cpp (hunk 4 above)
Comment (HowardHinnant) [inline src/ripple/app/ledger/impl/LedgerConsensus.cpp]: It depends on the implementation.  There could be a minor one if the implementation does a log(N) search for `end()`.  I think the standard forbids this, and libc++ doesn't.  Not positive about gcc and VS.

Glancing through this loop, this loop looks like a good opportunity to for a for-range:
````
for (auto const& i : closeTimes)
````
Doing so would also cache `end()`.

Comment (nbougalis) [inline src/ripple/app/ledger/impl/LedgerConsensus.cpp]: This really ought to be a ranged for.

File: src/beast/beast/module/core/native/linux_Files.cpp (hunk 5 above)
Comment (scottschurr) [inline src/beast/beast/module/core/native/linux_Files.cpp]: I don't care to get in an argument with you here, @vinniefalco , but when I look in the CodingStyle.md on develop (https://github.com/ripple/rippled/blob/develop/doc/CodingStyle.md) I see no such rule.  As a matter of fact this is the first time I've heard the rule you're suggesting.  I'm guessing you need to do a pull request for the style guide?

Comment (vinniefalco) [inline src/beast/beast/module/core/native/linux_Files.cpp]: @scottschurr the one in develop is a little bit behind, I'm using this:
https://github.com/vinniefalco/rippled/blob/style/doc/CodingStyle.md

File: src/beast/beast/module/core/native/linux_Files.cpp (hunk 7 above)
Comment (rec) [inline src/beast/beast/module/core/native/linux_Files.cpp]: Apparently the trailing underscore is correct, but I've been ordered to abandon this so...

A shame, I thought some of these seemed like legitimate issues.

**Comment by rec** [comment]
> Compiler implementations of shadow warnings are overly aggressive and create false positives (for example, in a lambda with no captures). 

Are you sure about this?  I managed to clean all the warnings except for that Boost one, which seems to be a legitimate complaint.

Anyway, I'll close this.

**Comment by vinniefalco** [comment]
Yes:
```
void foo()
{
  int i = 0;
  [](int i) {
    return i; }(i);
}
```
Produces a spurious shadowing warning.

**Comment by rec** [comment]
How is that spurious?

**Comment by vinniefalco** [comment]
@rec Because the lambda can't see the enclosing `i`.

**Comment by rec** [comment]
Pushed a new one.

I see that as a useful warning, honestly.  You have the same name with two different meanings in the same compilation unit.  It's confusing...!

**Comment by nbougalis** [comment]
I don't think this is a legitimate warning as far as shadowing is concerned: the scope of the lambda doesn't overlap with the scope of the function, so there's no way the two could shadow.

I personally don't find the situation with the lambda confusing. In fact, I'd argue that being able to give intuitive names to the arguments of the lambda - just like all arguments - is important in understanding its function.

There's some good stuff here (e.g. a `ScopedLock sl` shadowing an outer scoped log), but it also changes a lot of things that, in my opinion, don't need to be changed (e.g. the `DirectoryIterator` stuff over in Beast)

**Comment by HowardHinnant** [comment]
Nik, that could be dead-lock bait!

**Comment by rec** [comment]
Yes, I do want to bring that up.

I found several locks seemingly _inside_ other locks - surely that's just bad practice to be doing without noticing it?

I put comments each time I found it, could you take a look?  I should go back and detail exactly the order of the lock acquisition, too...

**Comment by HowardHinnant** [comment]
I'm having trouble finding the code...

File: src/ripple/app/ledger/impl/LedgerConsensus.cpp (hunk 4 above)
Comment (rec) [inline src/ripple/app/ledger/impl/LedgerConsensus.cpp]: Done.

**Comment by scottschurr** [comment]
Looks good to me.  :+1:

**Comment by rec** [comment]
Search for the string `WARNING:` in the diff to find the two places we have a lock inside a lock (that were found by this process).

**Comment by rec** [comment]
I just pushed a change relating only to the locks.  I removed one, and in the other two cases I commented them.  I think they're fine but have a look - if you agree I'll remove the scary comments...!

**Comment by rec** [comment]
PING!  I don't want to rebase etc. until we are sure that these locks aren't an issue, and we're sure that somewhere, we know that we're taking locks inside locks and have some way to guarantee we won't reverse the order somewhere else and cause a deadlock.

**Comment by nbougalis** [comment]
Are we fine with this? I don't like some of the gratuitous renames, but I can live with this and would rather not delay it.

**Comment by HowardHinnant** [comment]
No, the comments are incorrect in one place and there is a gratuitous recursive lock in another.

**Comment by rec** [comment]
All right!  I updated everything to Howard's latest, let me know when it's good and I'll fold it.

**Comment by rec** [comment]
All right, I rebased against develop and folded it...

**Comment by HowardHinnant** [comment]
:+1:

**Comment by scottschurr** [comment]
This looks good to me.  I noted a couple of nits that you may choose to address or not at your discretion.  I'm good either way.

:+1:

**Comment by vinniefalco** [comment]
Do any of these changes fix spurious warnings?

File: src/beast/beast/module/core/native/linux_Files.cpp (hunk 1 above)
Comment (vinniefalco) [inline src/beast/beast/module/core/native/linux_Files.cpp]: @scottschurr Your expectation is incorrect:
"The names of non-public data members and formal arguments used to initialize public data members have a trailing underscore."

File: src/ripple/app/ledger/impl/LedgerConsensus.cpp (hunk 2 above)
Comment (vinniefalco) [inline src/ripple/app/ledger/impl/LedgerConsensus.cpp]: When variables in a function are few, they should use "standard" names to avoid needless verbosity. If there were two or three loops running at once, then we might need a more descriptive name to distinguish them.

File: src/ripple/protocol/impl/STTx.cpp [hunk 10]
```diff
@@ -134,9 +134,9 @@ STTx::getMentionedAccounts () const
             if (std::find (accounts.cbegin (), accounts.cend (), na) == accounts.cend ())
                 accounts.push_back (na);
         }
-        else if (auto sa = dynamic_cast<STAmount const*> (&it))
+        else if (auto sa2 = dynamic_cast<STAmount const*> (&it))
         {
-            auto const& issuer = sa->getIssuer ();
+            auto const& issuer = sa2->getIssuer ();
```
Comment (vinniefalco) [inline src/ripple/protocol/impl/STTx.cpp]: Why was this changed? `sa2` is not in the same scope as `sa`.

**Comment by vinniefalco** [comment]
Eliminating warnings produced by `-Wshadow` is not a current objective. That *boost* does not consider these warnings important adds weight to the position that they are pointless. I see code being changed that has no problems. I am against the practice of renaming variables (`s2`, `hash2`, etc) only to satisfy a compiler warning that fixes no bug. Developers can try compiling with `-Wshadow` once in a while and fix the *real* errors that come up but we should not change working code.

File: src/beast/beast/module/core/native/linux_Files.cpp (hunk 7 above)
Comment (vinniefalco) [inline src/beast/beast/module/core/native/linux_Files.cpp]: @rec If there are legitimate issues (i.e. bugs) caused by shadowing, please address them in a separate pull request.
